lib/services/local/users.go

```diff
@@ -73,6 +74,186 @@ func (s *IdentityService) DeleteAllUsers(ctx context.Context) error {
+				if err != nil {
+					return nil, "", trace.Wrap(err)
+				}
+
+				out = append(out, user)
+				prevUser = name
+			}
+		}
+
+		// If there were no items returned that means there are either 0 users in the cluster
+		// or that we have exhausted the key range.
+		if len(result.Items) == 0 {
+			if len(out) == 0 {
+				return nil, "", nil
+			}
+
+			// If there are no more users its possible that the previous user being processed
+			// was never added to out because there was no additional user to transition to.
+			// Add the user from the collected items and return the full output.
+			if out[len(out)-1].GetName() != prevUser {
+				user, err := userFromUserItems(prevUser, users[prevUser])
+				if err != nil {
+					return nil, "", trace.Wrap(err)
+				}
+
+				out = append(out, user)
+			}
+
+			return out, "", nil
+		}
+
+		// If the number of users retrieved is more than the requested page size
+		// then return the number of users requested and the token for the user
+		// at page size + 1 in the slice.
+		if len(out) > pageSize {
+			nextKey := backend.GetPaginationKey(out[pageSize])
+			// Truncate the last item that was used to determine next row existence.
+			clear(out[pageSize:])
+			return out[:pageSize], nextKey, nil
+		}
+
+		// Since local auth secrets share the same prefix another batch of items from
+		// the backend may be necessary to fulfill the requested page. If the current
+		// start key matches the last item in the results then this page has already been
+		// retried once and means there are no more user entries to search so return the
+		// partial page of users.
+		nextStart := []byte(backend.NextPaginationKey(paginationKeyResource{name: string(result.Items[len(result.Items)-1].Key)}))
+		if bytes.Equal(rangeStart, nextStart) {
+			return out, "", nil
+		}
+
+		rangeStart = nextStart
+	}
+}
+
+// ListUsers returns the next page of users.
+func (s *IdentityService) ListUsers(ctx context.Context, pageSize int, nextToken string, withSecrets bool) ([]types.User, string, error) {
+	if withSecrets {
+		users, next, err := s.listUsersWitSecrets(ctx, pageSize, nextToken)
+		return users, next, trace.Wrap(err)
+	}
+
+	rangeStart := backend.Key(webPrefix, usersPrefix, nextToken)
+	rangeEnd := backend.RangeEnd(backend.ExactKey(webPrefix, usersPrefix))
+
+	// Adjust page size, so it can't be too large.
+	if pageSize <= 0 || pageSize > apidefaults.DefaultChunkSize {
+		pageSize = apidefaults.DefaultChunkSize
+	}
+
+	// Artificially inflate the limit to account for user secrets
+	// which have the same prefix.
+	limit := pageSize * 4
+
+	var out []types.User
+	for {
+		result, err := s.Backend.GetRange(ctx, rangeStart, rangeEnd, limit)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		// If there were no results just return an empty slice.
+		if len(result.Items) == 0 {
+			return out, "", nil
+		}
+
+		for _, item := range result.Items {
+			if !bytes.HasSuffix(item.Key, []byte(paramsPrefix)) {
```

Comment:
This was in the original `GetUsers` as well, but... are we sure that there's no way to get a MFA device in there that has an ID of `params`?